[2.5.0 Java] Cannot get session object in WebSocket controller method

### Play Version

2.5.0

### API

Java

### Operating System

Ubuntu 14.04
OS X 10.11.3

### JDK

Oracle 1.8.0_74

### Issue

The issue relates to the following controller code.

```java
public WebSocket notification() {
    Http.Session session = session();
    /* ... */
}
```

### Expected Behavior

Able to obtain session object without any errors.

### Actual Behavior

Internal Server Error with the following exception message:

```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[RuntimeException: There is no HTTP Context available from here.]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:269)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:195)
	at play.api.GlobalSettings$class.onError(GlobalSettings.scala:160)
	at play.api.DefaultGlobal$.onError(GlobalSettings.scala:188)
	at play.api.http.GlobalSettingsHttpErrorHandler.onServerError(HttpErrorHandler.scala:98)
	at play.core.server.netty.PlayRequestHandler$$anonfun$handle$1.applyOrElse(PlayRequestHandler.scala:128)
	at play.core.server.netty.PlayRequestHandler$$anonfun$handle$1.applyOrElse(PlayRequestHandler.scala:126)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:346)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:345)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
Caused by: java.lang.RuntimeException: There is no HTTP Context available from here.
	at play.mvc.Http$Context.current(Http.java:57)
	at play.mvc.Controller.session(Controller.java:90)
	at controllers.Api.notification(Api.java:775)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$23$$anonfun$apply$23.apply(Routes.scala:631)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$23$$anonfun$apply$23.apply(Routes.scala:631)
	at play.core.routing.HandlerInvokerFactory$$anon$16$$anon$17$$anonfun$call$2.apply(HandlerInvoker.scala:204)
	at play.core.routing.HandlerInvokerFactory$$anon$16$$anon$17$$anonfun$call$2.apply(HandlerInvoker.scala:203)
	at play.api.mvc.WebSocket$$anonfun$acceptOrResult$1.apply(WebSocket.scala:250)
	at play.api.mvc.WebSocket$$anonfun$acceptOrResult$1.apply(WebSocket.scala:249)
	at play.api.mvc.WebSocket$$anon$1.apply(WebSocket.scala:43)
```

Apparently, the ```Http.Context.current``` thread-local object is not set.

### Additional Comments

1. Standard HTTP controller methods can get the session object.
2. The same code for WebSocket controller methods can also get the session object in Play Framework 2.4.x.

--- a/framework/src/play-integration-test/src/test/java/play/it/http/websocket/WebSocketSpecJavaActions.java
+++ b/framework/src/play-integration-test/src/test/java/play/it/http/websocket/WebSocketSpecJavaActions.java
@@ -8,6 +8,7 @@ import akka.stream.javadsl.Keep;
 import akka.stream.javadsl.Sink;
 import akka.stream.javadsl.Source;
 import play.libs.F;
+import play.mvc.Http;
 import play.mvc.Results;
 import play.mvc.WebSocket;
 import scala.compat.java8.FutureConverters;
@@ -35,20 +36,26 @@ public class WebSocketSpecJavaActions {
     }
 
     public static WebSocket allowConsumingMessages(Promise<List<String>> messages) {
+        ensureContext();
         return WebSocket.Text.accept(request -> Flow.fromSinkAndSource(getChunks(messages::success), emptySource()));
     }
 
     public static WebSocket allowSendingMessages(List<String> messages) {
+        ensureContext();
         return WebSocket.Text.accept(request -> Flow.fromSinkAndSource(Sink.ignore(), Source.from(messages)));
     }
 
     public static WebSocket closeWhenTheConsumerIsDone() {
+        ensureContext();
         return WebSocket.Text.accept(request -> Flow.fromSinkAndSource(Sink.cancelled(), emptySource()));
     }
 
     public static WebSocket allowRejectingAWebSocketWithAResult(int statusCode) {
+        ensureContext();
         return WebSocket.Text.acceptOrResult(request -> CompletableFuture.completedFuture(F.Either.Left(Results.status(statusCode))));
     }
 
-
+    private static Http.Context ensureContext() {
+        return Http.Context.current();
+    }
 }

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/websocket/WebSocketSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/websocket/WebSocketSpec.scala
@@ -3,25 +3,28 @@
  */
 package play.it.http.websocket
 
+import java.net.URI
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.{Consumer, Function}
+
 import akka.actor._
 import akka.stream.scaladsl._
 import akka.util.ByteString
 import org.specs2.matcher.Matcher
+import play.api.Application
 import play.api.http.websocket._
 import play.api.inject.guice.GuiceApplicationBuilder
-import play.api.test._
-import play.api.Application
-import play.api.mvc.{ Handler, Results, WebSocket }
 import play.api.libs.iteratee._
+import play.api.mvc.{Handler, Results, WebSocket}
+import play.api.test._
 import play.core.routing.HandlerDef
 import play.it._
-import play.it.http.websocket.WebSocketClient.{ContinuationMessage, SimpleMessage, ExtendedMessage}
-import scala.concurrent.{ Future, Promise }
-import scala.concurrent.duration._
+import play.it.http.websocket.WebSocketClient.{ContinuationMessage, ExtendedMessage, SimpleMessage}
+
 import scala.concurrent.ExecutionContext.Implicits.global
-import java.net.URI
-import java.util.concurrent.atomic.AtomicReference
-import java.util.function.{ Consumer, Function }
+import scala.concurrent.duration._
+import scala.concurrent.{Future, Promise}
+import scala.reflect.ClassTag
 
 object NettyWebSocketSpec extends WebSocketSpec with NettyIntegrationSpecification
 object AkkaHttpWebSocketSpec extends WebSocketSpec with AkkaHttpIntegrationSpecification
@@ -412,15 +415,17 @@ trait WebSocketSpec extends PlaySpecification with WsTestClient with ServerInteg
 
     "allow handling a WebSocket in java" in {
 
+      import java.util.{List => JList}
+
       import play.core.routing.HandlerInvokerFactory
       import play.core.routing.HandlerInvokerFactory._
-      import java.util.{List => JList}
+
       import scala.collection.JavaConverters._
 
-      implicit def toHandler[J <: AnyRef](javaHandler: J)(implicit factory: HandlerInvokerFactory[J]): Handler = {
+      implicit def toHandler[J <: AnyRef](javaHandler: => J)(implicit factory: HandlerInvokerFactory[J], ct: ClassTag[J]): Handler = {
         val invoker = factory.createInvoker(
           javaHandler,
-          new HandlerDef(javaHandler.getClass.getClassLoader, "package", "controller", "method", Nil, "GET", "", "/stream")
+          new HandlerDef(ct.runtimeClass.getClassLoader, "package", "controller", "method", Nil, "GET", "", "/stream")
         )
         invoker.call(javaHandler)
       }
@@ -452,7 +457,7 @@ trait WebSocketSpec extends PlaySpecification with WsTestClient with ServerInteg
 
       import play.core.routing.HandlerInvokerFactory
       import play.core.routing.HandlerInvokerFactory._
-      import play.mvc.{ LegacyWebSocket, WebSocket => JWebSocket, Results => JResults }
+      import play.mvc.{LegacyWebSocket, Results => JResults, WebSocket => JWebSocket}
       import JWebSocket.{In, Out}
 
       implicit def toHandler[J <: AnyRef](javaHandler: J)(implicit factory: HandlerInvokerFactory[J]): Handler = {

--- a/framework/src/play/src/main/scala/play/core/routing/HandlerInvoker.scala
+++ b/framework/src/play/src/main/scala/play/core/routing/HandlerInvoker.scala
@@ -4,16 +4,16 @@
 package play.core.routing
 
 import java.util.Optional
-import java.util.concurrent.{ CompletableFuture, CompletionStage }
+import java.util.concurrent.{CompletableFuture, CompletionStage}
 
 import akka.stream.scaladsl.Flow
 import org.apache.commons.lang3.reflect.MethodUtils
 import play.api.mvc._
 import play.core.j
-import play.core.j.{ JavaHandlerComponents, JavaHandler, JavaActionAnnotations }
-import play.mvc.Http.RequestBody
+import play.core.j.{JavaActionAnnotations, JavaHandler, JavaHandlerComponents}
+import play.mvc.Http.{Context, RequestBody}
 
-import scala.compat.java8.{ OptionConverters, FutureConverters }
+import scala.compat.java8.{FutureConverters, OptionConverters}
 import scala.util.control.NonFatal
 
 /**
@@ -69,9 +69,9 @@ trait HandlerInvokerFactory[-T] {
 
 object HandlerInvokerFactory {
 
-  import play.mvc.{ Result => JResult, LegacyWebSocket, WebSocket => JWebSocket }
-  import play.core.j.JavaWebSocket
   import com.fasterxml.jackson.databind.JsonNode
+  import play.core.j.JavaWebSocket
+  import play.mvc.{LegacyWebSocket, Result => JResult, WebSocket => JWebSocket}
 
   private[routing] def handlerTags(handlerDef: HandlerDef): Map[String, String] = Map(
     play.api.routing.Router.Tags.RoutePattern -> handlerDef.path,
@@ -196,12 +196,25 @@ object HandlerInvokerFactory {
   }
 
   implicit def javaWebSocket: HandlerInvokerFactory[JWebSocket] = new HandlerInvokerFactory[JWebSocket] {
-    import play.http.websocket.{ Message => JMessage }
     import play.api.http.websocket._
     import play.api.libs.iteratee.Execution.Implicits.trampoline
+    import play.http.websocket.{Message => JMessage}
+
     def createInvoker(fakeCall: => JWebSocket, handlerDef: HandlerDef) = new HandlerInvoker[JWebSocket] {
       def call(call: => JWebSocket) = WebSocket.acceptOrResult[Message, Message] { request =>
-        FutureConverters.toScala(call(new j.RequestHeaderImpl(request))).map { resultOrFlow =>
+
+        val javaContext = JavaWebSocket.createJavaContext(request)
+
+        val callWithContext = {
+          try {
+            Context.current.set(javaContext)
+            FutureConverters.toScala(call(new j.RequestHeaderImpl(request)))
+          } finally {
+            Context.current.remove()
+          }
+        }
+
+        callWithContext.map { resultOrFlow =>
           if (resultOrFlow.left.isPresent) {
             Left(resultOrFlow.left.get.asScala())
           } else {

